### PR TITLE
pre load ondemand tablets for linear scans

### PR DESCRIPTION
**Pull request by keith-turner**
This PR contains my experimentation with making scans pre load ondemand tablets.  Below are some log messages from running BulkSplitOptimizationIT which creates around 57 splits in an on demand table and scans it.  The scan is faster that it was before this change, but still slower than if all the tablets were loaded.

```
2023-05-25T20:49:31,910 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 2 ondemand tablets to be hosted.
2023-05-25T20:49:37,046 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 2 ondemand tablets to be hosted.
2023-05-25T20:49:43,249 [logging.InternalLoggerFactory] DEBUG: Using SLF4J as the default logging framework
2023-05-25T20:49:43,308 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 4 ondemand tablets to be hosted.
2023-05-25T20:49:49,793 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 8 ondemand tablets to be hosted.
2023-05-25T20:49:55,886 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 11 ondemand tablets to be hosted.
2023-05-25T20:50:02,054 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 11 ondemand tablets to be hosted.
2023-05-25T20:50:09,089 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 11 ondemand tablets to be hosted.
2023-05-25T20:50:13,316 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 9 ondemand tablets to be hosted.
     100,000 records read |    2,369 records/sec |    7,900,000 bytes read |  187,168 bytes/sec | 42.208 secs   
```

These changes are not ready to commit, just posting in case anyone want to look over them.

### CODE REVIEW

**Comment by keith-turner** [comment]
Wondering if working on #3430 before continuing with this would be best.

**Comment by keith-turner** [comment]
With commit ddf749e seeing a big speedup when running the BulkSplitOptimizationIT

```
2023-05-26T21:39:06,004 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 2 ondemand tablets to be hosted.
2023-05-26T21:39:06,410 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 1 ondemand tablets to be hosted.
2023-05-26T21:39:06,471 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 1 ondemand tablets to be hosted.
2023-05-26T21:39:06,622 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 1 ondemand tablets to be hosted.
2023-05-26T21:39:06,665 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 1 ondemand tablets to be hosted.
2023-05-26T21:39:06,694 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 1 ondemand tablets to be hosted.
2023-05-26T21:39:06,721 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 1 ondemand tablets to be hosted.
2023-05-26T21:39:06,767 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 1 ondemand tablets to be hosted.
2023-05-26T21:39:06,903 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 2 ondemand tablets to be hosted.
2023-05-26T21:39:06,955 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 2 ondemand tablets to be hosted.
2023-05-26T21:39:07,052 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 3 ondemand tablets to be hosted.
2023-05-26T21:39:07,142 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 4 ondemand tablets to be hosted.
2023-05-26T21:39:07,268 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 5 ondemand tablets to be hosted.
2023-05-26T21:39:07,403 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 5 ondemand tablets to be hosted.
2023-05-26T21:39:07,540 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 5 ondemand tablets to be hosted.
2023-05-26T21:39:07,769 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 5 ondemand tablets to be hosted.
2023-05-26T21:39:07,936 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 5 ondemand tablets to be hosted.
2023-05-26T21:39:08,049 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 5 ondemand tablets to be hosted.
2023-05-26T21:39:08,165 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 5 ondemand tablets to be hosted.
2023-05-26T21:39:08,376 [clientImpl.ClientTabletCacheImpl] DEBUG: Requesting 3 ondemand tablets to be hosted.
     100,000 records read |   36,284 records/sec |    7,900,000 bytes read | 2,866,473 bytes/sec |  2.756 secs   
```

**Comment by keith-turner** [comment]
In #3437 processing of tablet hosting request was moved from the tserver to the manager. For #3437 a subset of the changes in this PR were taken and improved.  This PR still has client side changes that try to keep the next few tablets for a scan hosted.

With only the changes in #3437 that scan in BulkSplitOptimizationIT takes 17 seconds, which makes a hosting request for each of the ~50 tablets being scanned and waits.  The changes in the PR which have the changes in #3437 plus the changes to try to keep the next few tablets hosted only takes 2.7 seconds for the same scan of ~50 tablets.  So still need to continue experimenting with the changes in this PR.

**Comment by keith-turner** [comment]
I pulled some of the changes that were in the draft version of this PR into #3437 and #3438.  This PR now focuses solely on making linear scans host tablet ahead of time.  Those changes were forced pushed in 2a4b318.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java [hunk 1]
```diff
@@ -556,15 +557,68 @@ public CachedTablet findTablet(ClientContext context, Text row, boolean skipRow,
           String.format("%.3f secs", timer.scale(SECONDS)));
     }
 
-    if (tl != null && locationNeed == LocationNeed.REQUIRED && tl.getTserverLocation().isEmpty()) {
-      requestTabletHosting(context, List.of(tl));
-      return null;
+    if (tl != null && locationNeed == LocationNeed.REQUIRED) {
+      Map<KeyExtent,CachedTablet> extentsToHost =
+          findExtentsToHost(context, minimumHostAhead * 2, hostAheadRange, lcSession, tl);
```
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java:562]: why multiply hostAhead by 2 ?

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java [hunk 2]
```diff
@@ -556,15 +557,68 @@ public CachedTablet findTablet(ClientContext context, Text row, boolean skipRow,
...
+
+      var currTablet = extent;
+
+      for (int i = 0; i < hostAheadCount; i++) {
+        if (currTablet.endRow() == null || !hostAheadRange.contains(new Key(currTablet.endRow()))) {
+          break;
+        }
+
+        CachedTablet followingTablet = _findTablet(context, currTablet.endRow(), true, false, true,
+            lcSession, LocationNeed.REQUIRED);
```
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java:601]: the caller of this method requests tablet hosting for the result, won't REQUIRED bring the tablets online?

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java:601]: The method is only called when the calling method has location need of required.    I had thought of passing location need to this method and having it fail if it was not required.  I can try that and see what it looks like.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java]: I added the preconditions check which helps tightly couple the method with the caller.  Looking into the code the `_findTablet` method does not host tablets, it uses the location need to determine if it needs to do a metadata lookup for tablets in the cache w/o a location.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java (hunk 1 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java]: added a comment in 730a8eb

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java (hunk 2 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java]: added the check in 730a8eb

**Pull request merged**